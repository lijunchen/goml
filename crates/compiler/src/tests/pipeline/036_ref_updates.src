fn bump(cell: Ref[int]) -> int {
    let _ = cell := !cell + 1 in
    !cell
}

fn flip(flag: Ref[bool]) -> bool {
    let current = !flag in
    let _ = flag := !current in
    !flag
}

fn nested_total(cell: Ref[Ref[int]]) -> int {
    let inner = !cell in
    let before = !inner in
    let _ = inner := before + 2 in
    before + !inner
}

fn main() {
    let counter = ref 39 in
    let toggler = ref false in
    let nested = ref ref 3 in

    let bumped = bump(counter) in
    let flipped = flip(toggler) in
    let inner = !nested in
    let _ = inner := !inner + bumped in
    let nested_total_val = nested_total(nested) in

    let bool_check = !false in
    let _ = string_println(int_to_string(bumped + !counter)) in
    let _ = string_println(int_to_string(nested_total_val)) in
    let _ = string_println(bool_to_string(flipped && bool_check)) in
    ()
}
