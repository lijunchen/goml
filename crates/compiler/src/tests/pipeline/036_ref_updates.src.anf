fn bump(cell/0: ref int) -> int {
  let t8 = ref_get(cell/0) in
  let t7 = int_add(t8, 1) in
  let mtmp0 = ref_set(cell/0, t7) in
  ref_get(cell/0)
}

fn flip(flag/1: ref bool) -> bool {
  let current/2 = ref_get(flag/1) in
  let t9 = bool_not(current/2) in
  let mtmp1 = ref_set(flag/1, t9) in
  ref_get(flag/1)
}

fn nested_total(cell/3: ref ref int) -> int {
  let inner/4 = ref_get(cell/3) in
  let before/5 = ref_get(inner/4) in
  let t10 = int_add(before/5, 2) in
  let mtmp2 = ref_set(inner/4, t10) in
  let t11 = ref_get(inner/4) in
  int_add(before/5, t11)
}

fn main() -> unit {
  let counter/6 = ref_new(39) in
  let toggler/7 = ref_new(false) in
  let t12 = ref_new(3) in
  let nested/8 = ref_new(t12) in
  let bumped/9 = bump(counter/6) in
  let flipped/10 = flip(toggler/7) in
  let inner/11 = ref_get(nested/8) in
  let t14 = ref_get(inner/11) in
  let t13 = int_add(t14, bumped/9) in
  let mtmp3 = ref_set(inner/11, t13) in
  let nested_total_val/12 = nested_total(nested/8) in
  let bool_check/13 = bool_not(false) in
  let t17 = ref_get(counter/6) in
  let t16 = int_add(bumped/9, t17) in
  let t15 = int_to_string(t16) in
  let mtmp4 = string_println(t15) in
  let t18 = int_to_string(nested_total_val/12) in
  let mtmp5 = string_println(t18) in
  let t20 = bool_and(flipped/10, bool_check/13) in
  let t19 = bool_to_string(t20) in
  let mtmp6 = string_println(t19) in
  ()
}