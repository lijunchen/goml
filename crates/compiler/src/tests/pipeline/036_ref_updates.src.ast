fn main() {
    let reference = ref10 in
    let current = !reference in
    let _ = reference = current + 5 in
    let _ = string_println(int_to_string(!reference)) in
    ()
}

