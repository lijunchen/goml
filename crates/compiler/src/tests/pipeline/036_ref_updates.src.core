fn bump(cell/0: ref int) -> int {
  let mtmp0 = ref_set(cell/0, int_add(ref_get(cell/0), 1)) in
  ref_get(cell/0)
}

fn flip(flag/1: ref bool) -> bool {
  let current/2 = ref_get(flag/1) in
  let mtmp1 = ref_set(flag/1, bool_not(current/2)) in
  ref_get(flag/1)
}

fn nested_total(cell/3: ref ref int) -> int {
  let inner/4 = ref_get(cell/3) in
  let before/5 = ref_get(inner/4) in
  let mtmp2 = ref_set(inner/4, int_add(before/5, 2)) in
  int_add(before/5, ref_get(inner/4))
}

fn main() -> unit {
  let counter/6 = ref_new(39) in
  let toggler/7 = ref_new(false) in
  let nested/8 = ref_new(ref_new(3)) in
  let bumped/9 = bump(counter/6) in
  let flipped/10 = flip(toggler/7) in
  let inner/11 = ref_get(nested/8) in
  let mtmp3 = ref_set(inner/11, int_add(ref_get(inner/11), bumped/9)) in
  let nested_total_val/12 = nested_total(nested/8) in
  let bool_check/13 = bool_not(false) in
  let mtmp4 = string_println(int_to_string(int_add(bumped/9, ref_get(counter/6)))) in
  let mtmp5 = string_println(int_to_string(nested_total_val/12)) in
  let mtmp6 = string_println(bool_to_string(bool_and(flipped/10, bool_check/13))) in
  ()
}