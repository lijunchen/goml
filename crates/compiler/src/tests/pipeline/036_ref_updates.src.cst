FILE@0..179
  FN@0..179
    FnKeyword@0..2 "fn"
    Whitespace@2..3 " "
    Lident@3..7 "main"
    PARAM_LIST@7..10
      LParen@7..8 "("
      RParen@8..9 ")"
      Whitespace@9..10 " "
    BLOCK@10..179
      LBrace@10..11 "{"
      Whitespace@11..16 "\n    "
      EXPR_LET@16..177
        LetKeyword@16..19 "let"
        Whitespace@19..20 " "
        PATTERN_VARIABLE@20..30
          Lident@20..29 "reference"
          Whitespace@29..30 " "
        Eq@30..31 "="
        Whitespace@31..32 " "
        EXPR_LET_VALUE@32..39
          EXPR_PREFIX@32..39
            RefKeyword@32..35 "ref"
            Whitespace@35..36 " "
            EXPR_INT@36..39
              Int@36..38 "10"
              Whitespace@38..39 " "
        InKeyword@39..41 "in"
        Whitespace@41..46 "\n    "
        EXPR_LET_BODY@46..177
          EXPR_LET@46..177
            LetKeyword@46..49 "let"
            Whitespace@49..50 " "
            PATTERN_VARIABLE@50..58
              Lident@50..57 "current"
              Whitespace@57..58 " "
            Eq@58..59 "="
            Whitespace@59..60 " "
            EXPR_LET_VALUE@60..71
              EXPR_PREFIX@60..71
                Bang@60..61 "!"
                EXPR_LIDENT@61..71
                  Lident@61..70 "reference"
                  Whitespace@70..71 " "
            InKeyword@71..73 "in"
            Whitespace@73..78 "\n    "
            EXPR_LET_BODY@78..177
              EXPR_LET@78..177
                LetKeyword@78..81 "let"
                Whitespace@81..82 " "
                PATTERN_WILDCARD@82..84
                  WildcardKeyword@82..83 "_"
                  Whitespace@83..84 " "
                Eq@84..85 "="
                Whitespace@85..86 " "
                EXPR_LET_VALUE@86..110
                  EXPR_BINARY@86..110
                    EXPR_LIDENT@86..96
                      Lident@86..95 "reference"
                      Whitespace@95..96 " "
                    Eq@96..97 "="
                    Whitespace@97..98 " "
                    EXPR_BINARY@98..110
                      EXPR_LIDENT@98..106
                        Lident@98..105 "current"
                        Whitespace@105..106 " "
                      Plus@106..107 "+"
                      Whitespace@107..108 " "
                      EXPR_INT@108..110
                        Int@108..109 "5"
                        Whitespace@109..110 " "
                InKeyword@110..112 "in"
                Whitespace@112..117 "\n    "
                EXPR_LET_BODY@117..177
                  EXPR_LET@117..177
                    LetKeyword@117..120 "let"
                    Whitespace@120..121 " "
                    PATTERN_WILDCARD@121..123
                      WildcardKeyword@121..122 "_"
                      Whitespace@122..123 " "
                    Eq@123..124 "="
                    Whitespace@124..125 " "
                    EXPR_LET_VALUE@125..167
                      EXPR_CALL@125..167
                        EXPR_LIDENT@125..139
                          Lident@125..139 "string_println"
                        ARG_LIST@139..167
                          LParen@139..140 "("
                          ARG@140..165
                            EXPR_CALL@140..165
                              EXPR_LIDENT@140..153
                                Lident@140..153 "int_to_string"
                              ARG_LIST@153..165
                                LParen@153..154 "("
                                ARG@154..164
                                  EXPR_PREFIX@154..164
                                    Bang@154..155 "!"
                                    EXPR_LIDENT@155..164
                                      Lident@155..164 "reference"
                                RParen@164..165 ")"
                          RParen@165..166 ")"
                          Whitespace@166..167 " "
                    InKeyword@167..169 "in"
                    Whitespace@169..174 "\n    "
                    EXPR_LET_BODY@174..177
                      EXPR_UNIT@174..177
                        LParen@174..175 "("
                        RParen@175..176 ")"
                        Whitespace@176..177 "\n"
      RBrace@177..178 "}"
      Whitespace@178..179 "\n"