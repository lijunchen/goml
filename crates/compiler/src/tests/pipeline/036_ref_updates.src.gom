package main

import (
    "fmt"
)

func bool_to_string(x bool) string {
    if x {
        return "true"
    } else {
        return "false"
    }
}

func int_to_string(x int) string {
    return fmt.Sprintf("%d", x)
}

func bool_not(x bool) bool {
    return !x
}

func bool_and(x bool, y bool) bool {
    return x && y
}

func int_add(x int, y int) int {
    return x + y
}

func string_println(s string) struct{} {
    fmt.Println(s)
    return struct{}{}
}

func ref_new__Ref_int(value int) *int {
    return &value
}

func ref_get__Ref_int(reference *int) int {
    return *reference
}

func ref_set__Ref_int(reference *int, value int) struct{} {
    *reference = value
    return struct{}{}
}

func ref_new__Ref_bool(value bool) *bool {
    return &value
}

func ref_get__Ref_bool(reference *bool) bool {
    return *reference
}

func ref_set__Ref_bool(reference *bool, value bool) struct{} {
    *reference = value
    return struct{}{}
}

func ref_new__Ref_Ref_int(value *int) **int {
    return &value
}

func ref_get__Ref_Ref_int(reference **int) *int {
    return *reference
}

func bump(cell__0 *int) int {
    var ret38 int
    var t12 int = ref_get__Ref_int(cell__0)
    var t11 int = int_add(t12, 1)
    ref_set__Ref_int(cell__0, t11)
    ret38 = ref_get__Ref_int(cell__0)
    return ret38
}

func flip(flag__1 *bool) bool {
    var ret39 bool
    var current__2 bool = ref_get__Ref_bool(flag__1)
    var t13 bool = bool_not(current__2)
    ref_set__Ref_bool(flag__1, t13)
    ret39 = ref_get__Ref_bool(flag__1)
    return ret39
}

func nested_total(cell__3 **int) int {
    var ret40 int
    var inner__4 *int = ref_get__Ref_Ref_int(cell__3)
    var before__5 int = ref_get__Ref_int(inner__4)
    var t14 int = int_add(before__5, 2)
    ref_set__Ref_int(inner__4, t14)
    var t15 int = ref_get__Ref_int(inner__4)
    ret40 = int_add(before__5, t15)
    return ret40
}

func alias_bump(cell__6 *int) int {
    var ret41 int
    var alias__7 *int = cell__6
    var t17 int = ref_get__Ref_int(alias__7)
    var t16 int = int_add(t17, 5)
    ref_set__Ref_int(alias__7, t16)
    ret41 = ref_get__Ref_int(alias__7)
    return ret41
}

func pair_sum() int {
    var ret42 int
    var first__8 *int = ref_new__Ref_int(4)
    var second__9 *int = ref_new__Ref_int(6)
    var t19 int = ref_get__Ref_int(first__8)
    var t20 int = ref_get__Ref_int(second__9)
    var t18 int = int_add(t19, t20)
    ref_set__Ref_int(first__8, t18)
    var t21 int = ref_get__Ref_int(first__8)
    var t22 int = ref_get__Ref_int(second__9)
    ret42 = int_add(t21, t22)
    return ret42
}

func reassign_nested(nested__10 **int) int {
    var ret43 int
    var inner__11 *int = ref_get__Ref_Ref_int(nested__10)
    var t24 int = ref_get__Ref_int(inner__11)
    var t23 int = int_add(t24, 7)
    ref_set__Ref_int(inner__11, t23)
    ret43 = ref_get__Ref_int(inner__11)
    return ret43
}

func main0() struct{} {
    var ret44 struct{}
    var counter__12 *int = ref_new__Ref_int(39)
    var toggler__13 *bool = ref_new__Ref_bool(false)
    var t25 *int = ref_new__Ref_int(3)
    var nested__14 **int = ref_new__Ref_Ref_int(t25)
    var bumped__15 int = bump(counter__12)
    var flipped__16 bool = flip(toggler__13)
    var flipped_again__17 bool = flip(toggler__13)
    var inner__18 *int = ref_get__Ref_Ref_int(nested__14)
    var t27 int = ref_get__Ref_int(inner__18)
    var t26 int = int_add(t27, bumped__15)
    ref_set__Ref_int(inner__18, t26)
    var nested_total_val__19 int = nested_total(nested__14)
    var alias_total__20 int = alias_bump(counter__12)
    var pair_total__21 int = pair_sum()
    var reassigned__22 int = reassign_nested(nested__14)
    var bool_check__23 bool = bool_not(false)
    var t30 int = ref_get__Ref_int(counter__12)
    var t29 int = int_add(bumped__15, t30)
    var t28 string = int_to_string(t29)
    string_println(t28)
    var t33 int = int_add(nested_total_val__19, alias_total__20)
    var t32 int = int_add(t33, reassigned__22)
    var t31 string = int_to_string(t32)
    string_println(t31)
    var t34 string = int_to_string(pair_total__21)
    string_println(t34)
    var t37 bool = bool_and(flipped__16, flipped_again__17)
    var t36 bool = bool_and(t37, bool_check__23)
    var t35 string = bool_to_string(t36)
    string_println(t35)
    ret44 = struct{}{}
    return ret44
}

func main() {
    main0()
}
