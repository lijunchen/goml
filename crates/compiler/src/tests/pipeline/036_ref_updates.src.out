85
185
16
false
